src/qibo/backends/numpy.py

```diff
@@ -420,11 +422,25 @@ def execute_circuit(
                 for gate in circuit.queue:
                     state = gate.apply(self, state, nqubits)
 
-            if return_array:
-                return state
+            # set the circuit _final_state??
+
+            if circuit.has_unitary_channel:
+                # here we necessarily have `density_matrix=True`, otherwise
+                # execute_circuit_repeated would have been called
+                if circuit.measurements:
+                    circuit._final_state = CircuitResult(state, circuit, self, nshots)
```

Comment:
In the end I decided to keep it, as I think the circuit compilation breaks otherwise. Still, it would probably be better to change the name to `_final_result` as now even a stateless object could be returned in principle.

Reply:
Given that we may drop this completely in the future, I wouldn't worry much about the name. However we could open an issue to remember checking if this really needed and if yes maybe update the name in the future.